core/src/main/java/org/apache/iceberg/GenericManifestEntry.java

```diff
@@ -88,9 +93,19 @@ public Long snapshotId() {
     return snapshotId;
   }
 
+  /** @return the data sequence number of the file */
+  @Override
+  public Long dataSequenceNumber() {
+    return dataSequenceNumber;
+  }
+
   @Override
   public Long sequenceNumber() {
-    return sequenceNumber;
+    if (sequenceNumber != null) {
```

Comment:
I kept both fields for now as `sequenceNumber` may be set explicitly during inheritance.